In [30]:
from sklearn.neighbors import KNeighborsRegressor
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler,StandardScaler
import matplotlib.pyplot as plt

In [3]:
data=pd.read_csv('diamonds.csv')

In [5]:
data.head()

Unnamed: 0  carat      cut color clarity  depth  table  price     x     y  \
0           1   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98   
1           2   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84   
2           3   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07   
3           4   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23   
4           5   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35   

      z  
0  2.43  
1  2.31  
2  2.31  
3  2.63  
4  2.75

#### price: price in US dollars ( 326−− 18,823)
#### carat: weight of the diamond (0.2--5.01)
#### cut: quality of the cut (Fair, Good, Very Good, Premium, Ideal)
#### color: diamond colour, from J (worst) to D (best)
#### clarity: a measurement of how clear the diamond is (I1 (worst), SI2, SI1, VS2, VS1, VVS2, VVS1, IF (best))
#### x: length in mm (0--10.74)
#### y: width in mm (0--58.9)
#### z: depth in mm (0--31.8)
#### depth: total depth percentage = z / mean(x, y) = 2 * z / (x + y) (43--79)
#### table: width of top of diamond relative to widest point (43--95)

In [6]:
data=data.drop('Unnamed: 0',axis=1)

In [7]:
data

carat        cut color clarity  depth  table  price     x     y     z
0       0.23      Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1       0.21    Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2       0.23       Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3       0.29    Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4       0.31       Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75
...      ...        ...   ...     ...    ...    ...    ...   ...   ...   ...
53935   0.72      Ideal     D     SI1   60.8   57.0   2757  5.75  5.76  3.50
53936   0.72       Good     D     SI1   63.1   55.0   2757  5.69  5.75  3.61
53937   0.70  Very Good     D     SI1   62.8   60.0   2757  5.66  5.68  3.56
53938   0.86    Premium     H     SI2   61.0   58.0   2757  6.15  6.12  3.74
53939   0.75      Ideal     D     SI2   62.2   55.0   2757  5.83  5.87  3.64

[53940 rows x 10 columns]

In [9]:
data.dtypes

carat      float64
cut         object
color       object
clarity     object
depth      float64
table      float64
price        int64
x          float64
y          float64
z          float64
dtype: object

In [11]:
data.isnull().sum()

carat      0
cut        0
color      0
clarity    0
depth      0
table      0
price      0
x          0
y          0
z          0
dtype: int64

In [12]:
data.color.unique()

array(['E', 'I', 'J', 'H', 'F', 'G', 'D'], dtype=object)

In [13]:
data.cut.unique()

array(['Ideal', 'Premium', 'Good', 'Very Good', 'Fair'], dtype=object)

In [14]:
data.clarity.unique()

array(['SI2', 'SI1', 'VS1', 'VS2', 'VVS2', 'VVS1', 'I1', 'IF'],
      dtype=object)

In [17]:
cut={'Ideal':5,'Premium':4,'Good':3,'Very Good':2,'Fair':1}
color={'D':7,'E':6,'F':5,'G':4,'H':3,'I':2,'J':1}
clarity={'I1':1,'SI2':2,'SI1':3,'VS2':4,'VS1':5,'VVS2':6,'VVS1':7,'IF':8}

In [18]:
data['cut']=data['cut'].map(cut)

In [19]:
data['color']=data['color'].map(color)

In [20]:
data['clarity']=data['clarity'].map(clarity)

In [21]:
data.dtypes

carat      float64
cut          int64
color        int64
clarity      int64
depth      float64
table      float64
price        int64
x          float64
y          float64
z          float64
dtype: object

In [22]:
data.isnull().sum()

carat      0
cut        0
color      0
clarity    0
depth      0
table      0
price      0
x          0
y          0
z          0
dtype: int64

In [23]:
data

carat  cut  color  clarity  depth  table  price     x     y     z
0       0.23    5      6        2   61.5   55.0    326  3.95  3.98  2.43
1       0.21    4      6        3   59.8   61.0    326  3.89  3.84  2.31
2       0.23    3      6        5   56.9   65.0    327  4.05  4.07  2.31
3       0.29    4      2        4   62.4   58.0    334  4.20  4.23  2.63
4       0.31    3      1        2   63.3   58.0    335  4.34  4.35  2.75
...      ...  ...    ...      ...    ...    ...    ...   ...   ...   ...
53935   0.72    5      7        3   60.8   57.0   2757  5.75  5.76  3.50
53936   0.72    3      7        3   63.1   55.0   2757  5.69  5.75  3.61
53937   0.70    2      7        3   62.8   60.0   2757  5.66  5.68  3.56
53938   0.86    4      3        2   61.0   58.0   2757  6.15  6.12  3.74
53939   0.75    5      7        2   62.2   55.0   2757  5.83  5.87  3.64

[53940 rows x 10 columns]

In [25]:
x=data.drop('price',axis=1)
y=data['price']

In [26]:
x

carat  cut  color  clarity  depth  table     x     y     z
0       0.23    5      6        2   61.5   55.0  3.95  3.98  2.43
1       0.21    4      6        3   59.8   61.0  3.89  3.84  2.31
2       0.23    3      6        5   56.9   65.0  4.05  4.07  2.31
3       0.29    4      2        4   62.4   58.0  4.20  4.23  2.63
4       0.31    3      1        2   63.3   58.0  4.34  4.35  2.75
...      ...  ...    ...      ...    ...    ...   ...   ...   ...
53935   0.72    5      7        3   60.8   57.0  5.75  5.76  3.50
53936   0.72    3      7        3   63.1   55.0  5.69  5.75  3.61
53937   0.70    2      7        3   62.8   60.0  5.66  5.68  3.56
53938   0.86    4      3        2   61.0   58.0  6.15  6.12  3.74
53939   0.75    5      7        2   62.2   55.0  5.83  5.87  3.64

[53940 rows x 9 columns]

In [27]:
y

0         326
1         326
2         327
3         334
4         335
         ... 
53935    2757
53936    2757
53937    2757
53938    2757
53939    2757
Name: price, Length: 53940, dtype: int64

In [28]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.25,random_state=6)

In [31]:
model1=KNeighborsRegressor(n_neighbors=20)

In [32]:
model1.fit(x_train,y_train)

KNeighborsRegressor(n_neighbors=20)

In [33]:
model1.score(x_test,y_test)

0.9474057865182463

In [34]:
#Standard Scaler
scaler=StandardScaler()
x_train_scaled=scaler.fit_transform(x_train)
x_test_scaled=scaler.transform(x_test)

In [35]:
model2=KNeighborsRegressor(n_neighbors=20)

In [36]:
model2.fit(x_train_scaled,y_train)

KNeighborsRegressor(n_neighbors=20)

In [37]:
model2.score(x_test_scaled,y_test)

0.9654764529298038

In [39]:
#Minmax Scaler
scalermm=MinMaxScaler()
x_train_scaled=scalermm.fit_transform(x_train)
x_test_scaled=scalermm.transform(x_test)

In [40]:
model3=KNeighborsRegressor(n_neighbors=20)

In [41]:
model3.fit(x_train_scaled,y_train)

KNeighborsRegressor(n_neighbors=20)

In [42]:
model3.score(x_test_scaled,y_test)

0.9576242834933445